boinc service: optionally use a FHS environment

kierdavis opened the pull request:
This ensures that the BOINC project applications (immutable blobs) can
find the libraries they need to run, since many contain hardcoded paths.

Additionally, an option to install Virtualbox into the environment
and an option to install NVIDIA OpenCL/CUDA libraries into the
environment were added. Other OpenCL libraries could probably be
added similarly, but I don't have the hardware to test it on.

###### Motivation for this change


###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

kierdavis commented on nixos/modules/services/computing/boinc/client.nix:
```diff
@@ -6,6 +6,24 @@ let
+          ln -s $(realpath lib/vendors) etc/OpenCL/vendors
+          chmod -w etc
+        fi
+      '';
```
This feels like a horrible bodge. Is there a better way to do this?

abbradar replied:
Where does `lib/vendors` come from? We should probably install it to `etc/OpenCL/vendors` but that depends on what happens.

abbradar replied:
Eeh, strange. Github haven't shown me @zimbatm's previous message previously...

zimbatm replied:
It's my fault. I had started the review and then left it there. Your comment reminded that I had and just published without looking at your message. For some reason github gave me precendence.

kierdavis replied:
`lib/vendors` comes from the `nvidia-x11` package (and probably also the packages that provide the OpenCL implementations for AMD and Intel hardware too). It contains ICD files, which contain some information about how the generic OpenCL API is mapped to the device driver. Off the top of my head, I believe `nvidia-x11` provides one ICD file, `nvidia.icd`, which is necessary for the OpenCL API to detect NVIDIA GPUs.

`lib/vendors` is just where this ICD file is currently placed by the derivation of `nvidia-x11`. When I made this change I went for the less destructive option, but in my opinion `etc/OpenCL/vendors` would be the correct location for this file.

abbradar replied:
I'm not sure on OpenCL story but we should probably see how/if it works on NixOS currently, what paths are used etc. Probably we should have `nvidia-x11` installing it in the correct place.

BTW, you can make this less hacky by adding a bogus package with symlink to the environment:

```
pkg = runCommand "nvidia-opencl" {} ''
  mkdir -p $out/etc/OpenCL
  ln -s ${nvidia-x11}/lib/vendors $out/etc/OpenCL
'';

kierdavis replied:
Thanks for the tip, I've updated the PR to use something along those lines.

I'll try and have a look to see how much (if anything) will break if we change the location of `nvidia.icd`, and how much work it would be to change that.

abbradar replied:
I've already researched that a bit, it turns out OpenCL is broken on NixOS right now. Patch to fix it is,  luckily, trivial (and yes, includes using `/etc/OpenCL/vendors`; I'll make a PR soon.

abbradar replied:
See https://github.com/NixOS/nixpkgs/pull/21995.

abbradar replied:
I've merged the PR in question. `nvidia_x11` should be unnecesarry in the environment for OpenCL -- instead add `ocl-icd`.

kierdavis replied:
@abbradar awesome, thanks. It'll still need to optionally depend on `nvidia_x11` since the CUDA libraries are required as well for NVIDIA support, but I'll make OpenCL support a separate option that pulls in `ocl-icd`. This new option ought to cover AMD and Intel GPUs as well.

kierdavis committed:
boinc service: optionally use a FHS environment

This ensures that the BOINC project applications (immutable blobs) can
find the libraries they need to run.

Additionally, options to install VirtualBox, OpenCL and NVIDIA CUDA
into the environment were added.

kierdavis commented:
I've made these changes and rebased onto master. Hopefully CI should pass this time...

abbradar commented:
If I understand correctly different applications in BOINC can require any additional dependencies. Maybe instead of adding a flag for each one (which doesn't scale) you want to add `extraEnvPackages` or something like that? Just a suggestion, don't mind me if you think that's less convenient.

kierdavis commented:
Superseded by #30908, closing.
